infer/src/pulse/PulseExecutionDomain.ml

```diff
@@ -102,6 +108,170 @@ let pp fmt exec_state =
+  | _ ->
+      false
+
+
+let back_edge (prev : t list) (next : t list) (num_iters : int) : t list * int =
+  let has_infinite_state = find_infinite_state next in
+  if has_infinite_state then ([], -1)
+  else
+    let _ = num_iters in
+    let cfgnode = AnalysisState.get_node_exn () in
+    let same = phys_equal prev next in
+    let rec detect_elem e lst curi : bool * int =
+      match lst with
+      | [] ->
+          (false, -1)
+      | hd :: tl ->
+          if leq ~lhs:hd ~rhs:e then (true, curi) else detect_elem e tl (curi + 1)
+    in
+    let rec compute_workset prev next newset nextit : (t * int) list =
+      match next with
+      | [] ->
+          newset
+      | hd :: tl ->
+          let res, idx = detect_elem hd prev 0 in
+          let recidx = match idx with -1 -> nextit | _ -> idx in
+          if res then compute_workset prev tl newset (nextit + 1)
+          else [(hd, recidx)] @ compute_workset prev tl newset (nextit + 1)
+    in
+    let workset = compute_workset prev next [] 0 in
+    let prevlen = List.length prev in
+    let nextlen = List.length next in
+    let worklen = List.length workset in
+    let extract_pathcond hd : Formula.t =
+      match hd with
+      | AbortProgram summary ->
+          AbductiveDomain.Summary.get_path_condition summary
+      | ExitProgram summary ->
+          AbductiveDomain.Summary.get_path_condition summary
+      | LatentAbortProgram a ->
+          AbductiveDomain.Summary.get_path_condition a.astate
+      | LatentInvalidAccess a ->
+          AbductiveDomain.Summary.get_path_condition a.astate
+      | InfiniteProgram astate ->
+          AbductiveDomain.get_path_condition astate
+      | ExceptionRaised astate ->
+          AbductiveDomain.get_path_condition astate
+      | ContinueProgram astate ->
+          AbductiveDomain.get_path_condition astate
+      | LatentSpecializedTypeIssue {astate; _} ->
+          AbductiveDomain.Summary.get_path_condition astate
+    in
+    (* Used when the workset is not empty *)
+    let rec record_pathcond ws kl : int =
+      let cmp_four (a, b, c, d) (e, f, g, h) =
+        if phys_equal a e && phys_equal b f && phys_equal c g && phys_equal d h then true else false
+      in
+      match ws with
+      | [] ->
+          -1
+      | (hd, idx) :: tl -> (
+          let cond = extract_pathcond hd in
+          let pathcond = Formula.extract_path_cond cond in
+          let termcond = Formula.extract_terminal_atoms cond in
+          let termcond2 = Formula.extract_terminal_terms cond in
+          let key = (cfgnode, termcond, pathcond, termcond2) in
+          let dl_ws = get_widenstate () in
+          match dl_ws with
+          | None ->
+              -1
+          | Some ws -> (
+              let prevstate = Stdlib.Hashtbl.find_opt ws key in
+              match prevstate with
+              | None ->
+                  let _ = add_widenstate key in
+                  record_pathcond tl (kl @ [key])
```

Comment:
Adding at the end of the list is linear in `kl`, better to add `key` to the beginning? (and then reverse the list if the order matters)